eppo_core/src/event_ingestion/event.rs

```diff
@@ -1,11 +1,39 @@
 use crate::timestamp::Timestamp;
-use serde::{Deserialize, Serialize};
+use chrono::{DateTime, Utc};
+use serde::de::Deserializer;
+use serde::{Deserialize, Serialize, Serializer};
 
 #[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Hash)]
 pub(super) struct Event {
     pub uuid: uuid::Uuid,
+    #[serde(
+        serialize_with = "serialize_millis",
+        deserialize_with = "deserialize_millis"
+    )]
     pub timestamp: Timestamp,
     #[serde(rename = "type")]
     pub event_type: String,
     pub payload: serde_json::Value,
 }
+
+fn serialize_millis<S>(date: &Timestamp, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    // Convert to i64 representing milliseconds since the Unix epoch
+    let millis = date.timestamp_millis();
+    serializer.serialize_i64(millis)
+}
+
+fn deserialize_millis<'de, D>(deserializer: D) -> Result<Timestamp, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let millis = i64::deserialize(deserializer)?;
+    // Convert i64 millis back to Timestamp
+    let secs = millis / 1000;
+    let subsec_nanos = ((millis % 1000) * 1_000_000) as u32;
+    let naive = DateTime::from_timestamp(secs, subsec_nanos)
+        .ok_or_else(|| serde::de::Error::custom("Invalid timestamp"))?;
+    Ok(Timestamp::from(naive))
```

Comment:
There's a reverse method for `timestamp_millis`: [DateTime::from_timestamp_millis](https://docs.rs/chrono/latest/chrono/struct.DateTime.html#method.from_timestamp_millis) (I'm too lazy to verify the math here 😅)

`from_timestamp`/`from_timestamp_millis` also return a `DateTime<Utc>` (aka `Timestamp`), so the `Timestamp::from` on the list line is a no-op
```suggestion
    DateTime::from_timestamp_millis(millis)
      .ok_or_else(|| serde::de::Error::custom("Invalid timestamp"))
```